clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -0,0 +1,1021 @@
+         * @param callsToSend           A map of nodes to the calls they need to handle.
+         */
+        private void timeoutCallsToSend(long now, Map<Node, List<Call>> callsToSend) {
+            int numTimedOut = 0;
+            for (List<Call> callList : callsToSend.values()) {
+                numTimedOut += timeoutCalls(now, callList);
+            }
+            if (numTimedOut > 0)
+                log.debug("{}: timed out {} call(s) with assigned nodes.", clientId, numTimedOut);
+        }
+
+        /**
+         * Choose nodes for the RPCs in the newRpcs list.
+         *
+         * This function holds the lock for the minimum amount of time, to avoid blocking
+         * users of AdminClient who will also take the lock to add new RPC requests.
+         *
+         * @param now                   The current time in milliseconds.
+         * @param callsToSend           A map of nodes to the calls they need to handle.
+         *
+         * @return                      The new RPCs we need to process.
+         */
+        private void chooseNodesForNewCalls(long now, Map<Node, List<Call>> callsToSend) {
+            List<Call> newCallsToAdd = null;
+            synchronized (this) {
+                if (newCalls.isEmpty()) {
+                    return;
+                }
+                newCallsToAdd = newCalls;
+                newCalls = new LinkedList<>();
+            }
+            for (Call call : newCallsToAdd) {
+                chooseNodeForNewCall(now, callsToSend, call);
+            }
+        }
+
+        /**
+         * Choose a node for a new call.
+         *
+         * @param now                   The current time in milliseconds.
+         * @param callsToSend           A map of nodes to the RPCs they need to handle.
+         * @param call                  The call.
+         */
+        private void chooseNodeForNewCall(long now, Map<Node, List<Call>> callsToSend, Call call) {
+            Node node = call.nodeProvider.provide();
+            if (node == null) {
+                call.fail(now, new BrokerNotAvailableException(
+                    String.format("Error choosing node for %s: no node found.", call.callName)));
+                return;
+            }
+            log.trace("{}: assigned {} to {}", clientId, call, node);
+            getOrCreateListValue(callsToSend, node).add(call);
+        }
+
+        /**
+         * Send the RPCs which are ready.
+         *
+         * @param now                   The current time in milliseconds.
+         * @param callsToSend           The calls to send, by node.
+         * @param correlationIdToRpc    A map of correlation IDs to RPCs.
+         * @param callsInFlight         A map of nodes to the RPCs they have in flight.
+         *
+         * @return                      The minimum timeout we need for poll().
+         */
+        private long sendEligibleRpcs(long now, Map<Node, List<Call>> callsToSend,
```

Comment:
Generally, it's good to have consistent terminology to avoid confusion. The Kafka code typically just says `Requests` instead of `Rpcs`.